x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java

```diff
@@ -437,10 +452,15 @@ protected void onListShards(
                 delegate = new CCSSingleCoordinatorSearchProgressListener();
                 delegate.onListShards(shards, skipped, clusters, fetchPhase, timeProvider);
             }
-            searchResponse.compareAndSet(
-                null,
-                new MutableSearchResponse(shards.size() + skipped.size(), skipped.size(), clusters, threadPool.getThreadContext())
+            MutableSearchResponse mutableSearchResponse = new MutableSearchResponse(
+                shards.size() + skipped.size(),
+                skipped.size(),
+                clusters,
+                threadPool.getThreadContext()
             );
+            mutableSearchResponse.setSearchResponseMerger(searchResponseMerger); // MP TODO: should we just add this to the ctor?
```

Comment:
I was also going to leave a similar comment to yours (should we add this to the ctor?), but I left a comment below around using a supplier, that may actually help move the dependency to SearchResponseMerger as an argument of updatePartialResponse, perhaps. I have not thought about it a lot but it may make sense to consider that. An alternative could be a static method like `InternalAggregations#toplevelReduce` that takes all the things to be merged together as an argument.

Reply:
I may not be seeing the full picture of what you are suggesting (see my other comment about my confusion), but a few responses:

> move the dependency to SearchResponseMerger as an argument of updatePartialResponse

If this means to allow the SearchResponseMerger to hold/cache the results of the local cluster results of InternalAggs (or any results that come from the data sent to `MutableSearchResponse. updatePartialResponse`, we need to consider that we don't want to retain that result, because the actual full final result (for CCS MRT=true) comes in via CCSActionListener for the local cluster to the SearchResponseMerger (see my other comments for deeper dive into that).

In addition, I discovered a bug in this commit, where SearchResponseMerger was being set as `null` on the MutableSearchResponse when doing a CCS with remote clusters only (no local shards) because in that case onListShards is called before the call to `AsyncSearchTask.getSearchResponseMergerSupplier`, since the onListShards is called "manually" [here](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java#L363).

So in my follow-on commit, I have kept this as a setter, not part of the constructor, since I also need this code in `AsyncSearchTask.getSearchResponseMergerSupplier`:

```java
    @Override
    public Supplier<SearchResponseMerger> getSearchResponseMergerSupplier(
        SearchSourceBuilder source,
        TransportSearchAction.SearchTimeProvider timeProvider,
        AggregationReduceContext.Builder aggReduceContextBuilder
    ) {
        return () -> {
            searchResponseMerger = SearchTask.createSearchResponseMerger(source, timeProvider, aggReduceContextBuilder);
            MutableSearchResponse mutableSearchResponse = searchResponse.get();
            // when a remote-only CCS is done (no local shards), then onListShards is called BEFORE
            // getSearchResponseMergerSupplier, so check if the MutableSearchResponse has been created
            // and if so, set the SearchResponseMerger
            if (mutableSearchResponse != null) {
                mutableSearchResponse.setSearchResponseMerger(searchResponseMerger);
            }
            return searchResponseMerger;
        };
    }